### ISSUE TITLE
Failed to call RangeFromLengthLayer.get_out_data_from_opts

### ISSUE THREAD

**Issue by Atticus1806** [rwth-i6/returnn_common#134]
I am trying to do the following:

```
    l = nn.reduce(durations, mode="sum", axis=time_dim, use_time_mask=True)  # [B]
    t, _ = nn.range_from_length(l, out_spatial_dim=self.out_time_axis)  # [B,T]
```
with `durations` with shapes `[B,T']`

When getting the config this error occurs:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/u/hilmes/experiments/tts_new_sis/recipe/i6_private/users/hilmes/gauss_up.py", line 44, in test_upsampling
    out = upsample(inputs, durations, variances, time_dim)
  File "/u/hilmes/experiments/tts_new_sis/recipe/i6_private/users/hilmes/gauss_up.py", line 26, in __call__
    t, _ = nn.range_from_length(l, out_spatial_dim=self.out_time_axis)  # [B,T]
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn_common/nn/_generated_layers.py", line 653, in range_from_length
    layer = nn.make_layer({
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn_common/nn/base.py", line 635, in make_layer
    layer = Tensor(
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn_common/nn/base.py", line 112, in __init__
    data = _data_from_layer_dict(layer_dict, tensor=self)
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn_common/nn/base.py", line 815, in _data_from_layer_dict
    layer = net.construct_layer(net_dict=net_dict, name=out_name, add_layer=_add_layer)
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn/tf/network.py", line 961, in construct_layer
    return add_layer(name=name_with_prefix, layer_class=layer_class, **layer_desc)
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn_common/nn/base.py", line 811, in _add_layer
    raise ReturnnConstructTemplateException(msg) from exc
returnn_common.nn.base.ReturnnConstructTemplateException: Failed to call RangeFromLengthLayer.get_out_data_from_opts(
  out_spatial_dim=Dim{'upsamling_time'[B(-1)]},
  _network=<TFNetwork 'dummy_net' train>,
  _name='range_from_length',
  sources=[<InternalLayer 'reduce' out_type=Data{[B(-1)]}>],
  name='range_from_length',
  network=<TFNetwork 'dummy_net' train>,
)
```
I tried some debugging and I am not really understanding what is happening here, so I am afraid I cant give more background.

**Comment by albertz** [comment]
What is the exception from RASR? You forgot to add that here.

**Comment by albertz** [comment]
A simple reproducing test case would also be nice (similar as the other test cases).

**Comment by Atticus1806** [comment]
```
WARNING:tensorflow:From /u/hilmes/experiments/tts_new_sis/recipe/returnn/tf/network.py:477: calling Zeros.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
Traceback (most recent call last):
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn_common/nn/base.py", line 805, in _add_layer
    out_data = layer_class.get_out_data_from_opts(**layer_desc)
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn/tf/layers/basic.py", line 2817, in get_out_data_from_opts
    return Data(name="%s_output" % name, dim_tags=[dim_tag], dtype=dtype, sparse=sparse, dim=None)
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn/tf/util/data.py", line 2756, in __init__
    self.sanity_check(assume_complete=False)
  File "/u/hilmes/experiments/tts_new_sis/recipe/returnn/tf/util/data.py", line 2852, in sanity_check
    assert tag.dyn_size_ext.dtype in {"int32", "int64"}
AssertionError

The above exception was the direct cause of the following exception:
```
this is the rest of the message, but due to the last line of the message i assumed this was rather to be expected to happen.

**Comment by Atticus1806** [comment]
Its possible I messed something up, so here is my code:

```
class GaussianUpsampling(nn.Module):

  def __init__(self):
    super().__init__()

    self.out_time_axis = nn.SpatialDim("upsamling_time")

  def __call__(self, input: nn.Tensor, durations: nn.Tensor, variances: nn.Tensor, time_dim: nn.Dim):
    """
    Performs gaussian upsampling for given input, duration and (learned) variances
    :param input: H, [B, N, F]
    :param durations: d, [B, N]
    :param variances: [B, N]
    :param time_dim: Time dimension of the input, denoted by N
    :return: upsampled input
    """

    c = nn.cumsum(durations, axis=time_dim) - (0.5 * durations)  # [B,N]

    l = nn.reduce(durations, mode="sum", axis=time_dim, use_time_mask=True)  # [B]

    t, _ = nn.range_from_length(l, out_spatial_dim=self.out_time_axis)  # [B,T]

    w_t = -0.5 * (np.log(2.0 * np.pi) + nn.log(variances) + ((l - c) ** 2) / variances) # [B, T, N]
    w_t = nn.softmax(w_t, axis=time_dim)

    output = nn.dot(w_t, input, reduce=time_dim) # [B, T, F]

    return output

def test_upsampling():

  nn.reset_default_root_name_ctx()
  input_dim = nn.FeatureDim('input', 10)
  time_dim = nn.SpatialDim('input_time')
  inputs = nn.get_extern_data(nn.Data('input', dim_tags=[nn.batch_dim, time_dim, input_dim]))
  durations = nn.get_extern_data(nn.Data('durations', dim_tags=[nn.batch_dim, time_dim]))
  variances = nn.get_extern_data(nn.Data('variances', dim_tags=[nn.batch_dim, time_dim]))
  upsample = GaussianUpsampling()
  out = upsample(inputs, durations, variances, time_dim)
```

**Comment by Atticus1806** [comment]
Shapes etc. should be irellevant for now, they might have errors which I am trying to debug, this is why I want to generate the returnn config but it already crashes beforehand.

**Comment by albertz** [comment]
> this is the rest of the message, but due to the last line of the message i assumed this was rather to be expected to happen.

What do you mean? But this is the relevant error here. And you directly see the error there as well.

**Comment by albertz** [comment]
> A simpler case

I meant to make a test case just like the other test cases.

Your test is wrong also, you have the wrong dtype (range_from_length expects int32, just like the error tells you).

**Comment by albertz** [comment]
I just added a simple test:
```python

def test_range_from_length():
  # https://github.com/rwth-i6/returnn_common/issues/134
  class _Net(nn.Module):
    @nn.scoped
    def __call__(self, x: nn.Tensor, *, axis: nn.Dim) -> nn.Tensor:
      durations = nn.cast(nn.reduce(x, axis=x.feature_dim, mode="sum"), dtype="int32")
      durations.verify_out_shape({nn.batch_dim, axis})
      t, _ = nn.range_from_length(durations)
      return t

  net = _Net()
  config, net_dict = dummy_config_net_dict(net=net, with_axis=True)
  dummy_run_net(config, net=net)
```

Please next time prepare such a test yourself.

But anyway, it looks like you just have it wrong on your side. But I think this was pretty obvious from the error?

**Comment by Atticus1806** [comment]
> Please next time prepare such a test yourself.

Okay I will.

This solves my issue thanks and sorry for the work I caused. 
My thinking was that because of this line `The above exception was the direct cause of the following exception:`  the type error was not the root of the problem but rather caused by the call not working. Thats why I dismissed it, which is a mistake on my end I guess. Nevertheless I think the message is missleading.
